src/autoscaling/snapshotter.js

```diff
@@ -0,0 +1,170 @@
+            maxMemoryMbytes,
+        } = _.defaults(options, DEFAULT_OPTIONS);
+
+        checkParamOrThrow(eventLoopSnapshotIntervalSecs, 'options.eventLoopSnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(memorySnapshotIntervalSecs, 'options.memorySnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(snapshotHistorySecs, 'options.snapshotHistorySecs', 'Number');
+        checkParamOrThrow(maxBlockedRatio, 'options.maxBlockedRatio', 'Number');
+        checkParamOrThrow(minFreeMemoryRatio, 'options.minFreeMemoryRatio', 'Number');
+        checkParamOrThrow(maxMemoryMbytes, 'options.maxMemoryMbytes', 'Maybe Number');
+
+        this.eventLoopSnapshotIntervalMillis = eventLoopSnapshotIntervalSecs * 1000;
+        this.memorySnapshotIntervalMillis = memorySnapshotIntervalSecs * 1000;
+        this.snapshotHistoryMillis = snapshotHistorySecs * 1000;
+        this.maxBlockedMillis = 1000 * eventLoopSnapshotIntervalSecs * maxBlockedRatio;
+        this.minFreeMemoryRatio = minFreeMemoryRatio;
+        if (maxMemoryMbytes) this.maxMemoryBytes = maxMemoryMbytes * 1024 * 1024;
+
+        this.cpuSnapshots = [];
+        this.eventLoopSnapshots = [];
+        this.memorySnapshots = [];
+    }
+
+    async start() {
+        // Ensure max memory is correctly computed.
+        if (!this.maxMemoryBytes) {
+            const { totalBytes } = await getMemoryInfo();
+            if (isAtHome()) this.maxMemoryBytes = totalBytes;
+            else this.maxMemoryBytes = Math.ceil(totalBytes / 4);
+        }
+        // Add dummy snapshot to compare the first with.
+        this.eventLoopSnapshots.push({
+            createdAt: new Date(),
+            isOverloaded: false,
+            exceededMillis: 0,
+        });
+
+        // Start snapshotting.
+        this.eventLoopInterval = betterSetInterval(this._snapshotEventLoop.bind(this), this.eventLoopSnapshotIntervalMillis);
+        this.memoryInterval = betterSetInterval(this._snapshotMemory.bind(this), this.memorySnapshotIntervalMillis);
+        if (isAtHome()) events.on(ACTOR_EVENT_NAMES.CPU_INFO, this._snapshotCpu.bind(this));
+    }
+
+    async stop() {
+        betterClearInterval(this.eventLoopInterval);
+        betterClearInterval(this.memoryInterval);
+        if (isAtHome()) events.removeListener(ACTOR_EVENT_NAMES.CPU_INFO, this._snapshotCpu);
+        // Allow event loop to unwind before stop returns.
+        await new Promise(resolve => setImmediate(resolve));
+    }
+
+    getMemorySample(sampleDurationMillis) {
+        return this._getSample(this.memorySnapshots, sampleDurationMillis);
+    }
+
+    getEventLoopSample(sampleDurationMillis) {
+        return this._getSample(this.eventLoopSnapshots, sampleDurationMillis);
+    }
+
+    getCpuSample(sampleDurationMillis) {
+        return this._getSample(this.cpuSnapshots, sampleDurationMillis);
+    }
+
+    _getSample(snapshots, sampleDurationMillis) { // eslint-disable-line class-methods-use-this
```

Comment:
Do we need `sampleDurationMillis` as we either want to know all or last item? It just adds extra logic.

Reply:
I don' know if last item is enough. We could get into a cycle where we'd overload, skip running a task, get fine and then run a task that would immediately overload again... It feels like using a sample of at least 2 - 5 secs could smooth it out.